src/Context/CachedContextFactory.php

```diff
@@ -0,0 +1,31 @@
+<?php
+
+namespace Roave\BetterReflection\Context;
+
+use Roave\BetterReflection\Context\ContextFactory;
+
+class CachedContextFactory implements ContextFactory
+{
+    private $contexts = [];
```

Comment:
I would stop at the in-memory caching as-is here. Further caching is not needed in my opinion, and can live in consumer projects instead. I also wouldn't want to deal with serialization here - let someone else have a painful experience.